src/core/initialize.rs

```diff
@@ -80,40 +80,90 @@ impl StackTraceGetter {
+                             Ok(remoteprocess::Error::NixError(e)) => match e {
+                                 nix::Error::Sys(nix::errno::Errno::EPERM) => {
+                                     debug!("EPERM for process {} : {:?}", self.process.pid, e);
+                                     return Err(MemoryCopyError::PermissionError);
+                                 }
+                                 nix::Error::Sys(nix::errno::Errno::ESRCH) => {
+                                     match Process::new(self.process.pid) {
+                                         Ok(_) => {
+                                             debug!("Process {} is actually alive, spurious ESRCH detected", self.process.pid);
+                                             return Err(MemoryCopyError::ProcessNotLocked);
+                                         },
+                                         Err(e) => {
+                                             debug!("ESRCH Process {} is not alive: {:?}", self.process.pid, e);
+                                             return Err(MemoryCopyError::ProcessEnded);
+                                         }
+                                     }
+                                 }
+                                 _ => {}
+                             },
+                             _ => {}
+                         }
+                     }
+                     debug!("Unhandled error {:?}", e);
+                     Err(e.into())
+                 })
+        };
+
         let _lock;
         if self.lock_process {
-            _lock = self
-                .process
-                .lock()
-                .context("locking process during stack trace retrieval")
-                .or_else(|e| {
-                    if let Some(source) = e.source() {
-                        match source.downcast_ref::<remoteprocess::Error>().ok_or(source) {
-                            Ok(remoteprocess::Error::IOError(e)) => {
-                                match e.kind() {
-                                    std::io::ErrorKind::NotFound => {
+            let mut retries = 20;
+            loop {
+                match try_lock(&self.process) {
+                    Ok(l) => {
+                        debug!(
+                            "Was able to lock process {}, {} retries remained ",
+                            self.process.pid, retries
+                        );
+                        _lock = l;
+                        break;
+                    }
+                    Err(e) => match e {
+                        MemoryCopyError::ProcessNotLocked | MemoryCopyError::PermissionError => {
+                            if retries == 0 {
+                                debug!("Exhausted retries to lock process");
+                                match e {
+                                    MemoryCopyError::PermissionError => {
                                         return Err(MemoryCopyError::ProcessEnded)
```

Comment:
This is a bit awkward, but is to mimic existing behavior as to not break the `test_spawn_record_children_subprocesses` test, but I believe that this is actually a bug in the other teast, which was previously swallowing permission errors, treating them as though the program had exited successfully.

Were it not for this, we could treat EPERM as a ProcessNotLocked error, but we need to distinguish between the two to preserve compatibility.

We should probably file a separate bug for this EPERM leading to a 0 exit code that has also been uncovered by this test case.